### PR TITLE
classy-afrovenator: Add billing tab to /settings

### PR DESCRIPTION
**Pull request by taravancil**
## Links
* [classy-afrovenator.glitch.me](https://classy-afrovenator.glitch.me)
* [[ch6209]](https://app.clubhouse.io/glitch/story/6209/add-billing-tab-to-settings): Add billing tab to /settings

## GIF/Screenshots:
**To come. The backend is returning `500`s for the `/glitchPro` endpoints. Will update when Justin is in and can investigate.**

## Changes:
This adds a "Subscription" tab to /settings behind the `pufferfish` feature flag.

## How To Test:
*Note that this is the dogfood implementation of this feature, so review should be focused on "Does it work well enough to test internally and iterate on?" rather than polish and cleverness.*

### Testing Subscription Tab Visibility
To see the tab:

1. Log in
2. Make sure you're in [@testing-team](https://glitch.com/@testing-team)
3. Make sure the `pufferfish` flag is enabled on [classy-afrovenator.glitch.me/secret](https://classy-afrovenator.glitch.me/secret)

| Case  | Expected Outcome | 
| ------------- | ------------- |
| In Testing Team + Enabled the feature on /secret  | Subscription tab visible |
| In Testing Team + Disabled the feature on /secret  | Subscription tab NOT visible |
| NOT In Testing Team + Enabled the feature on /secret  | Subscription NOT tab visible  |
| NOT In Testing Team + Disabled the feature on /secret  | Subscription NOT tab visible |

### Testing Creating + Cancelling a Subscription
**To come. The backend is returning `500`s for the `/glitchPro` endpoints. Will update when Justin is in and can investigate.**

## Feedback I'm looking for:
- Can you create a subscription with test data?
- Can you cancel that subscription?
- Do the `useStripe` and `useSubscription` hooks feel ok to you?

### CODE REVIEW

File: src/state/stripe.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import useScript from 'Hooks/use-script';
+
+export default function useStripe() {
+  const [stripe, setStripe] = useState(null);
+  const stripeJS = 'https://js.stripe.com/v3/';
+
+  const [loaded] = useScript(stripeJS);
+  useEffect(() => {
+    if (loaded) {
+      setStripe(window.Stripe('pk_test_WIJxKl0T1xT8zSb2StHu8zlo'));
```
Comment (taravancil) [inline src/state/stripe.js]: Just noting that this is a publishable key. 😄

File: src/components/account-settings-overlay/subscription-settings.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { useAPIHandlers } from 'State/api';
+
+function SubscriptionSettings() {
+  const stripe = useStripe();
+  const subscriptionStatus = useSubscriptionStatus();
+  const { createSubscriptionSession, cancelSubscription } = useAPIHandlers();
+
+  async function subscribe() {
+    try {
+      const { data } = await createSubscriptionSession({ successUrl: 'https://glitch.com', cancelUrl: 'https://glitch.com/pricing' });
```
Comment (sarahzinger) [inline src/components/account-settings-overlay/subscription-settings.js]: apologies if you're already on it, but just in case I think these urls don't feel quite right, yeah? Not 100% sure what we should do for dogfood, we could keep em on the pricing page with some kind of flash of success? Or redirect them back to their user profile so they see the new collection?

Comment (taravancil) [reply src/components/account-settings-overlay/subscription-settings.js]: Yeah, I just put those as placeholders until we have proper mockups and stuff. Does that feel ok to you?

File: src/hooks/use-script.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    for (const script of scripts) {
+      if (script.src) {
+        loadedScripts.add(script.src);
+      }
+    }
+
+    const script = document.createElement('script');
+
+    if (loadedScripts.has(src)) {
+      setLoaded(true);
```
Comment (sarahzinger) [inline src/hooks/use-script.js]: could you not have a script tag with a src that wasn't loaded? like it errored or something? Then again this is probably 💯 for dogfood imo!

Comment (taravancil) [reply src/hooks/use-script.js]: Yes 😝 In fact if you simulate a slow network you will probably see how this fails. But for dogfood I think it's fine.

File: src/presenters/pages/settings.js [hunk 4]
```diff
@@ -15,35 +18,75 @@ import styles from './settings.styl';
...
+        <Tabs selectedIndex={currentTab} onSelect={(tabIndex) => setCurrentTab(tabIndex)}>
+          <div className={styles.settingsPage}>
+            <TabList className={styles.settingsActions}>
+              <Tab key="account" className={styles.settingsTab}>
+                <Button as="span">Account</Button>
+              </Tab>
+              <Tab key="subscription" hidden={!showSubscriptionTab} className={styles.settingsTab}>
+                <Button as="span">Subscription</Button>
+              </Tab>
+              {/* <Button disabled onClick>Privacy & Notifications</Button> */}
```
Comment (sarahzinger) [inline src/presenters/pages/settings.js]: I think we can kill this comment now btw

File: src/components/account-settings-overlay/subscription-settings.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      {subscriptionStatus.isActive ? (
+        <>
+          <Text defaultMargin>Subscribed to the Extra Memory monthly plan for $14 per month.</Text>
+          <Button disabled={!stripe} variant="secondary" onClick={cancel}>
+            Cancel Subscription
+          </Button>
+        </>
+      ) : (
+        <>
+          <Text defaultMargin>Subscribe to the Extra Memory monthly plan for $14 per month.</Text>
```
Comment (whimsicallyson) [inline src/components/account-settings-overlay/subscription-settings.js]: when I click through to Stripe checkout, it says $10/mo?

Comment (taravancil) [reply src/components/account-settings-overlay/subscription-settings.js]: This is all placeholder content for the dogfood stage :)

Comment (whimsicallyson) [reply src/components/account-settings-overlay/subscription-settings.js]: fair!

File: src/presenters/pages/settings.js [hunk 6]
```diff
@@ -15,35 +18,74 @@ import styles from './settings.styl';
...
-            {/* <Button disabled onClick>Privacy & Notifications</Button> */}
+        <Heading tagName="h1">
+          Settings <Icon className={emoji} icon="key" />
+        </Heading>
+        <Tabs selectedIndex={currentTab} onSelect={(tabIndex) => setCurrentTab(tabIndex)}>
+          <div className={styles.settingsPage}>
+            <TabList className={styles.settingsActions}>
+              <Tab key="account" className={styles.settingsTab}>
+                <Button as="span">Account</Button>
+              </Tab>
```
Comment (whimsicallyson) [inline src/presenters/pages/settings.js]: I want more whitespace between these buttons, but I get that this is dogfooding and that's not a priority. just stating for the record :)

Comment (taravancil) [reply src/presenters/pages/settings.js]: Yep 😄 We don't have proper mockups yet, so everything is a bit ugly and more about testing Stripe and such.

File: src/hooks/use-script.js (hunk 3 above)
Comment (sarahzinger) [reply src/hooks/use-script.js]: cool maybe we should just leave a comment so we come back to it later

File: src/components/account-settings-overlay/subscription-settings.js (hunk 2 above)
Comment (sarahzinger) [reply src/components/account-settings-overlay/subscription-settings.js]: yeah that's fine! Can I vote for now we go to the user page if it succeeds and stay on the billing page if it fails?

**Comment by taravancil** [comment]
@sarahzinger thanks! I'm working on tests, but will let you know when it's ready.

**Comment by taravancil** [comment]
@sarahzinger this is ready for another round of review!

Changed since your last review:

- Refactored the settings page to use a `<SettingsTabsContainer>` and `<SettingsTabs>` component, for the sake of making it possible to test the tabs by passing in props
- Updated the Stripe `successUrl` and  `cancelUrl`
- Added a note about the `useScript` edge case

File: test/components/settings-tabs.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        <MockContext currentUser={this.currentUser} location={'settings'}>
+          <SettingsTabs userPasswordEnabled />
+        </MockContext>,
+      );
+      expect(
+        wrapper
+          .find(SettingsTabs)
+          .find('li[role="tab"]')
+          .at(1)
+          .prop('hidden'),
```
Comment (sarahzinger) [inline test/components/settings-tabs.js]: I think if we ever update the tab order this test will no longer be testing the right thing yeah? Would it make sense to add a data attribute to make it easier to test the tabs?
